cpp/src/arrow/compute/api_scalar.h

```diff
@@ -450,5 +450,145 @@ ARROW_EXPORT
+///
+/// \param[in] values input to extract day of year from
+/// \param[in] ctx the function execution context, optional
+/// \return the resulting datum
+///
+/// \since 4.0.0
+/// \note API not yet finalized
+ARROW_EXPORT Result<Datum> DayOfYear(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief DayOfWeek returns day of the week value for each element of `values`
+///
+/// \param[in] values input to extract dat of the week from
+/// \param[in] ctx the function execution context, optional
+/// \return the resulting datum
+///
+/// \since 4.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> DayOfWeek(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Hour returns hour value for each element of `values`
+///
+/// \param[in] values input to extract hour from
+/// \param[in] ctx the function execution context, optional
+/// \return the resulting datum
+///
+/// \since 4.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> Hour(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Minute returns minutes value for each element of `values`
+///
+/// \param[in] values input to extract minutes from
+/// \param[in] ctx the function execution context, optional
+/// \return the resulting datum
+///
+/// \since 4.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> Minute(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Second returns seconds value for each element of `values`
+///
+/// \param[in] values input to extract seconds from
+/// \param[in] ctx the function execution context, optional
+/// \return the resulting datum
+///
+/// \since 4.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> Second(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Millisecond returns milliseconds value for each element of `values`
+///
+/// \param[in] values input to extract milliseconds from
+/// \param[in] ctx the function execution context, optional
+/// \return the resulting datum
+///
+/// \since 4.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> Millisecond(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Microsecond returns microseconds value for each element of `values`
+///
+/// \param[in] values input to extract microseconds from
+/// \param[in] ctx the function execution context, optional
+/// \return the resulting datum
+///
+/// \since 4.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> Microsecond(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Nanosecond returns nanoseconds value for each element of `values`
+///
+/// \param[in] values input to extract nanoseconds from
+/// \param[in] ctx the function execution context, optional
+/// \return the resulting datum
+///
+/// \since 4.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> Nanosecond(const Datum& values, ExecContext* ctx = NULLPTR);
```

Comment:
Are this the total nanoseconds, or only the nanoseconds part after the microseconds? (i.e. are microseconds and nanoseconds additive components, or rather both representing the full fractional part of the second using a different resolution?)

Reply:
Seeing the tests, it's clearly the first (additive components), which I think is good. I am only wondering if we can make that clearer in the terminology of the docstring

Reply:
Oh yeah indeed. This would behave in the latter way which would be surprising to users. Will fix.

Reply:
To be clear, judging the tests, the current implementation is not using the "total" nanoseconds right now, but only the part after the microseconds. Personally I think this is the useful behaviour

Reply:
Do you know by hear what does pandas do? [Documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.microsecond.html) is not clear.
Perhaps it would be good to then have a `subseconds` function that would return the total nanoseconds since midnight?

Reply:
Yeah, I was also looking at pandas to compare ;) And that was the reason I am asking.

So the situation is a bit complex there:

```
>>> ts = pd.Timestamp("2012-01-02 03:04:05.123456789")
>>> ts.second
5
```

There is a `microsecond` attribute which gives the total microseconds, but this is inherited from `datetime.datetime` (as this actually stores the sub-second part as microseconds):

```
>>> ts.microsecond
123456
```

and then there is also a `nanosecond` attribute which is only the part after the microseconds (so the "additive componenent", not total nanoseconds):
```
>>> ts.nanosecond
789
```


(the example in the pandas docs is indeed not good! As you can't tell which of the two cases it is because of the used value)

Reply:
Well this is not ideal :).
I propose to implement this as `total units since midnight` and then add an exception in python for nanoseconds.

Reply:
We don't need to match exactly what pandas does to be clear. pyarrow.compute should just follow Arrow's behaviour.

So for Arrow the question is, what is most useful: the total units vs the component? (or do we need both?)

As comparison, the R lubridate package only has component getter functions up to `second()`, and not for `milli/micro/..second`, but the `second` function returns the fractional part of the second as well:

```R
> library(lubridate)
> x <- ymd_hms("2012-03-26 01:02:03.123456")
> second(x)
[1] 3.123456
```

Reply:
How about for `2012-03-26 01:02:03.123456789`:
* subsecond -> 123456789
* second[int] -> 3
* second[float] -> 3.123456789
* microsecond -> 123
* millisecond -> 456
* nanosecond -> 789

This way we have minimal `postprocessing` in languages other than c++